fields/sql.go

```diff
@@ -0,0 +1,106 @@
+package fields
+
+import (
+	"database/sql/driver"
+	"encoding"
+	"encoding/json"
+	"reflect"
+)
+
+// Valuer fulfills the sql/driver.Valuer interface which deserializes our
+// struct field value into a valid SQL value.
+type Valuer struct {
+	*Descriptor
+	value reflect.Value
+}
+
+// Value satisfies the sql/driver.Valuer interface.
+// The value should be one of the following:
+//    int64
+//    float64
+//    bool
+//    []byte
+//    string
+//    time.Time
+//    nil - for NULL values
+func (f Valuer) Value() (driver.Value, error) {
+	// Ideally, all the interfaces below would be implemented as normal methods, not indirect
+	// methods. However, in order to be safe we take the address of pointers where we can.
+	// This specifically supports protobuf.
+	i := f.value.Interface()
+
+	// If our interface supports driver.Valuer we can immediately short-circuit as this is what the
+	// MySQL driver would do.
+	if valuer, ok := i.(driver.Valuer); ok {
+		return valuer.Value()
+	}
+
+	// Return early if the value is nil. Ideally we would do a `i == nil` comparison here, but
+	// unfortunately for us, `nil` is typed and that would always return false.
+	if f.value.Kind() == reflect.Ptr && f.value.IsNil() {
+		return nil, nil
+	}
+
+	// Override serialization behavior with tags (these take precedence over how a type would
+	// usually be serialized).
+	// Example:
+	// struct {
+	//   Blob proto.Blob `sql:",binary"`       // ensures that Marshal or MarshalBinary is used.
+	//   IP IP `sql:",string"`                 // ensures that its MarshalText method
+	//	                                       // is used for serialization.
+	//   IP IP `sql:",text"`                   // is equivalent to the former field
+	//   JSON map[string]string `sql:",json"`  // ensures that json.Marshal is used on the value.
+	// }
+	switch {
+	case f.Tags.Contains("binary"):
+		if iface, ok := i.(marshaler); ok {
+			return iface.Marshal()
+		}
+		if iface, ok := i.(encoding.BinaryMarshaler); ok {
+			return iface.MarshalBinary()
+		}
+	case f.Tags.Contains("string"), f.Tags.Contains("text"):
+		if iface, ok := i.(encoding.TextMarshaler); ok {
+			return iface.MarshalText()
+		}
+	case f.Tags.Contains("json"):
+		if b, ok := i.([]byte); ok {
+			return b, nil
+		}
```

Comment:
why would one use the `json` tag if the value is a string? i.e. there is nothing for the marshaler to do

Reply:
Removed